android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/MainActivity.java

```diff
@@ -607,10 +607,16 @@ private void downloadKMP(String scheme) {
     try {
       String url = data.toString();
       if (url != null) {
-        // Create filename by extracting packageID from the URL
-        String urlNoQuery = url.substring(0, url.indexOf(data.getQuery())-1);
-        String filename = FileUtils.getFilename(urlNoQuery) + FileUtils.KEYMANPACKAGE;
 
+        String filename = "";
+        if (FileUtils.isKeymanLink(url)) {
+          // Create filename by extracting packageID from the Keyman download keyboard URL
+          String urlNoQuery = url.substring(0, url.indexOf(data.getQuery()) - 1);
+          filename = FileUtils.getFilename(urlNoQuery) + FileUtils.KEYMANPACKAGE;
+        } else {
+          // Otherwise, filename should be at the end of the URL
+          filename = FileUtils.getFilename(url);
```

Comment:
It is still possible that there will be a query `?...` appended to the URL (perhaps even a search `#...`). We need to make sure we work just from the path component. In which case, the only difference between the two code paths here is the first one appends `FileUtils.KEYMANPACKAGE`.